### PR TITLE
feat(dflash): selectable attention backend (sdpa/eager) with dense mask

### PR DESCRIPTION
**Pull request by Sawyer117**
## What this does
  Lets you pick the attention backend for the DFlash draft via `--draft-attn-impl`.
  The default doesn't change (still flex) — this just adds `sdpa`/`eager` for cases
  where flex attention can't run.

  ## Why
  On Ascend NPU, `flex_attention` flat-out errors (`FlexAttention is only supported on
  CUDA, CPU or HPU devices`, #531), and DFlash hardcodes flex — so there was no way to
  train it on NPU. The `@torch.compile`'d forward is also unhappy on NPU, and flex +
  compile has been flaky on CUDA too (#544). So I added a path that doesn't need flex.

  ## How
  - Added `build_anchor_block_dense_mask` — it builds the same anchor-block mask that
    `create_anchor_block_mask_mod` already builds, just as a plain dense boolean tensor
    instead of a flex `BlockMask`. Same rules (a query sees context before its anchor +
    its own block, sliding-window supported), with a unit test checking it matches the
    flex version exactly.
  - Added a `--draft-attn-impl {simple_flex_attention,sdpa,eager}` flag. Default is
    flex, so CUDA behavior is unchanged. `sdpa`/`eager` use the dense mask instead of a
    flex `BlockMask`.
  - With `sdpa`, the op becomes a fused FlashAttention kernel on NPU, so the full
    attention scores never get materialized — `max_anchors` can stay large.

  ## Tested
  - Unit test: dense mask matches the flex `mask_mod` exactly (full attention + sliding
    window).
  - End-to-end: trained Qwen3-8B DFlash on Ascend (CANN 9.0.0 / vllm-ascend 0.20.2rc1)
    with `--draft-attn-impl sdpa` — runs, loss ~1.3, position_1 acc ~0.40 and climbing.

  ## Notes
  - flex is block-sparse and faster when it works, so it stays the default; sdpa/eager
    is the fallback for when flex isn't available.
  - EAGLE3 and PEAGLE hardcode flex the same way; this PR is DFlash-only, but the same
    approach drops in for them as a follow-up.

### CODE REVIEW

File: scripts/train.py [hunk 1]
```diff
@@ -726,6 +726,16 @@ def parse_args():
         default=256,
         help="Maximum anchor positions for DFlash training (default: 256)",
     )
+    parser.add_argument(
+        "--draft-attn-impl",
+        type=str,
+        default=None,
```
Comment (menogrey) [inline scripts/train.py]: default="simple_flex_attention" may be better?

File: src/speculators/models/dflash/core.py [hunk 2]
```diff
@@ -205,30 +226,51 @@ def _build_attention_mask(self, loss_mask, lengths, device):
...
-            mask_mod, q_len, kv_len = create_anchor_block_mask_mod(
-                lengths=lengths.to(device),
-                total_seq_len=total_seq_len,
-                anchor_positions=anchor_positions,
-                block_size=self.block_size,
-                sliding_window=None,
-            )
-            full_attn_mask = create_block_mask(
-                mask_mod, B=None, H=None, Q_LEN=q_len, KV_LEN=kv_len, device=device
-            )
```
Comment (fynnsu) [inline src/speculators/models/dflash/core.py:217]: I think we can use:
```python
from torch.nn.attention.flex_attention import create_mask, create_block_mask

# in __init__
self._create_mask_fn = create_block_mask if self._attn_impl == "simple_flex_attention" else create_mask

# in forward
mask_mod, q_len, kv_len = create_anchor_block_mask_mod(
    lengths=lengths.to(device),
    total_seq_len=total_seq_len,
    anchor_positions=anchor_positions,
    block_size=self.block_size,
    sliding_window=None,
)

full_attn_mask = self._create_mask_fn(mask_mod, B=None, H=None, Q_LEN=q_len, KV_LEN=kv_len, device=device)
```

The `create_mask` function from pytorch creates a dense mask from the given mask_mod function and has pretty much the same signature as `create_block_mask`.

File: src/speculators/models/dflash/core.py [hunk 3]
```diff
@@ -205,30 +226,51 @@ def _build_attention_mask(self, loss_mask, lengths, device):
...
+            else:
+                mask_mod, q_len, kv_len = create_anchor_block_mask_mod(
+                    lengths=lengths.to(device),
+                    total_seq_len=total_seq_len,
+                    anchor_positions=anchor_positions,
+                    block_size=self.block_size,
+                    sliding_window=self.sliding_window,
+                    sliding_window_non_causal=self.sliding_window_non_causal,
+                )
+                sliding_window_attn_mask = create_block_mask(
```
Comment (fynnsu) [inline src/speculators/models/dflash/core.py]: Same deal with sliding window.

File: src/speculators/models/dflash/core.py [hunk 4]
```diff
@@ -144,6 +158,13 @@ def from_training_args(
             kwargs["verifier_name_or_path"],
         )
 
+        # Attention backend for the draft layers (--draft-attn-impl); defaults to
+        # flex in __init__. Set 'sdpa'/'eager' on backends without flex_attention
+        # (e.g. Ascend NPU).
+        draft_attn_impl = kwargs.get("draft_attn_impl")
```
Comment (fynnsu) [inline src/speculators/models/dflash/core.py]: Can be simplified if we just set a valid default value for `--draft-attn-impl`

File: src/speculators/models/dflash/core.py [hunk 5]
```diff
@@ -44,11 +47,22 @@ def __init__(
...
-        # Forcibly override config settings
+        """Build the DFlash draft model from its config.
+
+        Resolves the draft attention backend (`config.transformer_layer_config.
+        _attn_implementation`): flex by default, or the dense-mask `sdpa`/`eager`
+        path on backends without `flex_attention` (e.g. Ascend NPU).
+        """
+        # Default to flex attention. On backends without flex_attention (e.g. Ascend
+        # NPU), pass --draft-attn-impl sdpa (or eager) to use a materialized dense
+        # mask instead; see build_anchor_block_dense_mask.
```
Comment (fynnsu) [inline src/speculators/models/dflash/core.py]: Remove

File: scripts/train.py (hunk 1 above)
Comment (fynnsu) [reply scripts/train.py]: Agreed

**Review by fynnsu — CHANGES_REQUESTED** [review]
Added some comments. This is worth adding but I think we can simplify this implementation a bit. Please see notes below. 

Also, it would be great if you could add this to the other algorithms (eagle3, peagle, mtp), using the approach described below? But I can also do that in a follow up if you don't have time.

**Comment by Sawyer117** [comment]
> Added some comments. This is worth adding but I think we can simplify this implementation a bit. Please see notes below.
> 
> Also, it would be great if you could add this to the other algorithms (eagle3, peagle, mtp), using the approach described below? But I can also do that in a follow up if you don't have time.

Hi @fynnsu thank you very much for the reply! I agree with all the comments and will fix them tomorrow, meanwhile if you want to fix it for eagle3/peagle/mtp, feel free to go ahead🙏

**Comment by fynnsu** [comment]
> I agree with all the comments and will fix them tomorrow, meanwhile if you want to fix it for eagle3/peagle/mtp, feel free to go ahead🙏

Sounds good! I'll add this as a follow up, once this pr lands.

File: scripts/train.py (hunk 1 above)
Comment (Sawyer117) [reply scripts/train.py]: totally agree, just fixed.

File: src/speculators/models/dflash/core.py (hunk 2 above)
Comment (Sawyer117) [reply src/speculators/models/dflash/core.py:217]: much appreicated,   I followed this pattern, but wrapped the final call in `_create_attention_mask()` so the same logic can be reused for both full attention and sliding-window attention.

  `self._create_mask_fn` is still the backend selector:
  - `simple_flex_attention` -> `create_block_mask`
  - `sdpa` / `eager` -> `create_mask`

  Then `_create_attention_mask()` builds `mask_mod, q_len, kv_len` and calls `self._create_mask_fn(...)`. So it’s the same structure as your suggestion, just factored into a helper to avoid duplicating the block for the sliding-window path.

File: src/speculators/models/dflash/core.py (hunk 3 above)
Comment (Sawyer117) [reply src/speculators/models/dflash/core.py]: fixed as mentioned above

File: src/speculators/models/dflash/core.py (hunk 4 above)
Comment (Sawyer117) [reply src/speculators/models/dflash/core.py]: just fixed

File: src/speculators/models/dflash/core.py (hunk 5 above)
Comment (Sawyer117) [reply src/speculators/models/dflash/core.py]: done

**Comment by Sawyer117** [comment]
Hi @fynnsu anything else from my side that needs to be done for the PR to be merged?

**Review by fynnsu — APPROVED** [review]

**Comment by fynnsu** [comment]
> Hi @fynnsu anything else from my side that needs to be done for the PR to be merged?

No, I think you should be good. I just pushed a couple extra tests, and should be almost ready to merge.
